linalg/slepc.cpp

```diff
@@ -0,0 +1,342 @@
+// CONTRIBUTING.md for details.
+
+#include "../config/config.hpp"
+
+#ifdef MFEM_USE_MPI
+#ifdef MFEM_USE_PETSC
+#ifdef MFEM_USE_SLEPC
+
+#include "linalg.hpp"
+
+#include "slepc.h"
+
+// Error handling
+// Prints SLEPc's stacktrace and then calls MFEM_ABORT
+// We cannot use PETSc's CHKERRQ since it returns a PetscErrorCode
+#define PCHKERRQ(obj,err) do {                                                   \
+     if ((err))                                                                  \
+     {                                                                           \
+        PetscError(PetscObjectComm((PetscObject)(obj)),__LINE__,_MFEM_FUNC_NAME, \
+                   __FILE__,(err),PETSC_ERROR_REPEAT,NULL);                      \
+        MFEM_ABORT("Error in SLEPc. See stacktrace above.");                     \
+     }                                                                           \
+  } while(0);
+#define CCHKERRQ(comm,err) do {                                \
+     if ((err))                                                \
+     {                                                         \
+        PetscError(comm,__LINE__,_MFEM_FUNC_NAME,              \
+                   __FILE__,(err),PETSC_ERROR_REPEAT,NULL);    \
+        MFEM_ABORT("Error in SLEPc. See stacktrace above.");   \
+     }                                                         \
+  } while(0);
+
+
+static PetscErrorCode ierr;
+
+namespace mfem
+{
+void MFEMInitializeSlepc()
+{
+   MFEMInitializeSlepc(NULL,NULL,NULL,NULL);
+}
+
+void MFEMInitializeSlepc(int *argc,char*** argv)
+{
+   MFEMInitializeSlepc(argc,argv,NULL,NULL);
+}
+
+void MFEMInitializeSlepc(int *argc,char ***argv,const char rc_file[],
+                         const char help[])
+{
+   ierr = SlepcInitialize(argc,argv,rc_file,help);
+   MFEM_VERIFY(!ierr,"Unable to initialize SLEPc");
+}
+
+void MFEMFinalizeSlepc()
+{
+   ierr = SlepcFinalize();
+   MFEM_VERIFY(!ierr,"Unable to finalize SLEPc");
+}
+
+
+SlepcEigenSolver::SlepcEigenSolver(MPI_Comm comm, const std::string &prefix,
+                                   bool wrap)
+{
+   clcustom = false;
+   _tol = PETSC_DEFAULT;
+   _max_its = PETSC_DEFAULT;
+   _num_conv = -1;
+   _wrap = wrap;
+   VR = NULL;
+   VC = NULL;
+   operatorset = false;
+
+   ierr = EPSCreate(comm,&eps); CCHKERRQ(comm,ierr);
+   ierr = EPSSetOptionsPrefix(eps, prefix.c_str()); PCHKERRQ(eps, ierr);
+}
+
+SlepcEigenSolver::~SlepcEigenSolver()
+{
+   MPI_Comm comm;
+   ierr = PetscObjectGetComm((PetscObject)eps,&comm); PCHKERRQ(eps,ierr);
+   ierr = EPSDestroy(&eps); CCHKERRQ(comm,ierr);
+}
+
+
+void SlepcEigenSolver::SetOperator(const Operator &op)
+{
+   PetscParMatrix *pA = const_cast<PetscParMatrix *>
+                        (dynamic_cast<const PetscParMatrix *>(&op));
+   const HypreParMatrix *hA = dynamic_cast<const HypreParMatrix *>(&op);
+   const Operator       *oA = dynamic_cast<const Operator *>(&op);
+   bool delete_pA = false;
+
+   if (!pA)
```

Comment:
Since SLEPc is based on PETSc, I don't see the reason why we need to convert here.
It is responsibility of the users to pass a `PetscParMatrix`. There are plenty of ways to construct it, and it should be done by users, not by MFEM. I suggest you change the interface with `const PetscParMatrix&` as argument. 
This is different for the linear and nonlinear solvers, since you want to have a similar API for MFEM and PETSc based solvers. @v-dobrev and @tzanio ?

Reply:
I can do that. In comparison, the Hypre LOBPCG takes `Operator&` as argument and casts to a `HypreParMatrix`, but I am not really trying to match exactly the LOBPCG interface.

Reply:
I agree with @stefanozampini. The PETSc interface gives the user more flexibility to decide whether he wants to pass a nested or monolithic matrix etc.

Reply:
I changed the interface to only take `PetscParMatrix`, and changed ex11p accordingly.